extensions-core/lookups-cached-global/src/main/java/io/druid/server/lookup/namespace/cache/CacheScheduler.java

```diff
@@ -0,0 +1,487 @@
+    private final ExtractionNamespaceCacheFactory<T> cachePopulator;
+    private final AtomicLong updateCounter = new AtomicLong(0);
+    private final CountDownLatch firstRunLatch = new CountDownLatch(1);
+    private final CountDownLatch startLatch = new CountDownLatch(1);
+
+    private EntryImpl(final T namespace, final Entry<T> entry, final ExtractionNamespaceCacheFactory<T> cachePopulator)
+    {
+      try {
+        this.namespace = namespace;
+        this.updaterFuture = schedule(namespace);
+        this.entryCleaner = createCleaner(entry);
+        this.cachePopulator = cachePopulator;
+        activeEntries.incrementAndGet();
+      }
+      finally {
+        startLatch.countDown();
+      }
+    }
+
+    private Cleaner createCleaner(Entry<T> entry)
+    {
+      return Cleaner.create(entry, new Runnable()
+      {
+        @Override
+        public void run()
+        {
+          disposeFromCleaner();
+        }
+      });
+    }
+
+    private Future<?> schedule(final T namespace)
+    {
+      final long updateMs = namespace.getPollMs();
+      Runnable command = new Runnable()
+      {
+        @Override
+        public void run()
+        {
+          updateCache();
+        }
+      };
+      if (updateMs > 0) {
+        return cacheManager.scheduledExecutorService().scheduleAtFixedRate(command, 0, updateMs, TimeUnit.MILLISECONDS);
+      } else {
+        return cacheManager.scheduledExecutorService().schedule(command, 0, TimeUnit.MILLISECONDS);
+      }
+    }
+
+    private void updateCache()
+    {
+      try {
+        // Ensures visibility of the whole EntryImpl's state (fields and their state).
+        startLatch.await();
+
+        CacheState currentCacheState = cacheStateHolder.get();
+        if (!Thread.currentThread().isInterrupted() && currentCacheState != NoCache.ENTRY_CLOSED) {
+          final String currentVersion = currentVersionOrNull(currentCacheState);
+          final VersionedCache newVersionedCache =
+              cachePopulator.populateCache(namespace, entryId, currentVersion, CacheScheduler.this);
+          if (newVersionedCache != null) {
+            swapCacheState(newVersionedCache);
+          } else {
+            log.debug("%s: Version `%s` not updated, the cache is not updated", entryId, currentVersion);
+          }
+        }
+      }
+      catch (Throwable t) {
+        try {
+          dispose();
+          log.error(t, "Failed to update %s", entryId);
+        }
+        catch (Exception e) {
+          t.addSuppressed(e);
+        }
+        if (Thread.currentThread().isInterrupted() || t instanceof Error) {
+          throw Throwables.propagate(t);
+        }
+      }
+    }
+
+    private String currentVersionOrNull(CacheState currentCacheState)
+    {
+      if (currentCacheState instanceof VersionedCache) {
+        return ((VersionedCache) currentCacheState).version;
+      } else {
+        // currentCacheState == CACHE_NOT_INITIALIZED
```

Comment:
This method is called only from one place, from if block with condition `currentCacheState  != ENTRY_CLOSED`. However it shouldn't matter so I removed this comment as confusing.